### PR TITLE
ENH: Download Checkpoints across AML workspaces

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Enable checkpoints transfer across AML workspaces

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from datetime import datetime, timedelta
+from health_azure import get_workspace
+from health_ml.utils.common_utils import DEFAULT_AML_CHECKPOINT_DIR
+from azureml.core import Workspace
+
+
+def get_ssl_checkpoint_url(
+    run_id: str,
+    checkpoint_filename: str,
+    expiry_hours: int = 1,
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py:11]: You might want to increase that, 1 hour is really very short. 1 week default?

File: hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--run_id', type=str, help='The run id of the SSL model checkpoint')
+    parser.add_argument('--checkpoint_filename', type=str, default='last.ckpt',
+                        help='The filename of the SSL model checkpoint. Default: last.ckpt')
+    parser.add_argument(
+        '--expiry_hours', type=int, default=1, help='The number of hours for which the SAS token is valid. Default: 1'
+    )
+    args = parser.parse_args()
+    args.run_id = '1f391509-f0a7-41d9-bef5-06713739fb0b'
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py:51]: hardcoded run_id? debug code?

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 3]
```diff
@@ -60,11 +64,58 @@ class EncoderParams(param.Parameterized):
     encoding_chunk_size: int = param.Integer(
         default=0, doc="If > 0 performs encoding in chunks, by enconding_chunk_size tiles " "per chunk"
     )
+    ssl_checkpoint: str = param.String(
+        default="", doc="Optional run id from which to load checkpoint if using SSLEncoder"
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:68]: this is no longer a run ID?

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 4]
```diff
@@ -60,11 +64,58 @@ class EncoderParams(param.Parameterized):
...
+
+    @property
+    def ssl_checkpoint_is_aml_run_id(self) -> bool:
+        return checkpoint_is_aml_run_id(self.ssl_checkpoint)
+
+    @property
+    def is_valid_src_checkpoint(self) -> bool:
+        if self.ssl_checkpoint:
+            return self.ssl_checkpoint_is_local_file or self.ssl_checkpoint_is_url or self.ssl_checkpoint_is_aml_run_id
+        return True
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:87]: This is copy/pasted from src_checkpoint, can we do better?
Provide a class that does the checking of a string that describes a checkpoint, and if OK, does the download?

File: hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py (hunk 2 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/scripts/generate_ssl_checkpoint_url.py:51]: Good catch, left over from debugging.

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py (hunk 3 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:68]: Yes, will update the doc

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py (hunk 4 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:87]: Great suggestion, will make it happen!

File: hi-ml/src/health_ml/utils/checkpoint_utils.py [hunk 5]
```diff
@@ -232,3 +233,111 @@ def download_checkpoint_if_necessary(self) -> None:
...
+        :param checkpoints_folder: The checkpoints folder to which the checkpoint should be downloaded if it is a URL or
+            AzureML run ID.
+        :raises ValueError: If the checkpoint is not a local file, URL or AzureML run ID.
+        :raises FileNotFoundError: If the checkpoint is a URL or AzureML run ID and the download fails.
+        :return: The path to the checkpoint file.
+        """
+        if self.is_local_file:
+            checkpoint_path = Path(self.checkpoint)
+        elif self.is_url:
+            download_folder = checkpoints_folder / MODEL_WEIGHTS_DIR_NAME
```
Comment (ant0nsc) [inline hi-ml/src/health_ml/utils/checkpoint_utils.py:331]: There's some inconsistency here - if the input is a run ID, we download directly into the checkpoint folder, but if it's a URL we add this special folder. Why?

File: hi-ml/src/health_ml/utils/checkpoint_utils.py [hunk 6]
```diff
@@ -232,3 +233,111 @@ def download_checkpoint_if_necessary(self) -> None:
...
+            checkpoint_path = self.download_weights(url=self.checkpoint, download_folder=download_folder)
+        elif self.is_aml_run_id:
+            downloader = CheckpointDownloader(run_id=self.checkpoint, download_dir=checkpoints_folder)
+            downloader.download_checkpoint_if_necessary()
+            checkpoint_path = downloader.local_checkpoint_path
+        else:
+            raise ValueError("Unable to determine how to get the checkpoint path.")
+
+        if checkpoint_path is None or not checkpoint_path.is_file():
+            raise FileNotFoundError(f"Could not find the weights file at {checkpoint_path}")
```
Comment (ant0nsc) [inline hi-ml/src/health_ml/utils/checkpoint_utils.py:342]:
```suggestion
            raise FileNotFoundError(f"Could not find the file at {checkpoint_path}")
```

**Review by ant0nsc — APPROVED** [review]
Looking great, just a couple of small-fry comments. Consider a separate doc page about how to download checkpoints etc - if we at some point really add keyvault tokens, it will be easier to add that there.

File: hi-ml-cpath/environment.yml [hunk 7]
```diff
@@ -81,6 +81,7 @@ dependencies:
     - azure-mgmt-keyvault==10.0.0
     - azure-mgmt-resource==21.1.0
     - azure-mgmt-storage==20.0.0
+    - azure-storage-blob==12.5.0
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/environment.yml:84]: Yes, it is the outcome of lock_env

File: hi-ml/src/health_ml/utils/checkpoint_utils.py (hunk 5 above)
Comment (kenza-bouzid) [reply hi-ml/src/health_ml/utils/checkpoint_utils.py:331]: When we DL from AML runs we add the run_id folder (see CheckpointDownloader) similarly we add a custom folder for  weights from URLs

**Pull request merged**